Fix the children order in line graph and various refactorings

ibbem opened the pull request:
The line graph exporter tries to merge unchanged children. However, when a child was moved, it cannot be classified as an unchaged edge when linearized (same problems as representing moves in line diffs). This resulted in a bug that makes the line graph exporter non-injective. This was fixed in 3a778b52c15e00345cf413c49ce1cb0ca7085097.

All other commits in this PR are just refactorings I had lying around. All of these commits should probably be reviewed independently but I'm too lazy and thus I'm opening only one PR 😅

ibbem committed:
refactor: eliminate superfluous `try`

ibbem committed:
test: fix the DiffTreeParser tests for non-GumTree matchers

Previously, all tests would run with the default GumTree matcher instead
of the matcher stated in the test case file name.

As stated in my bachelor thesis, some matchers fail to return a valid
matching. Hence, some test cases need to be removed. In particular, the
theta matcher contains a bug that results in matchings that are
inconsistent (e.g., A->B but B->C). Furthermore, the
gumtree-partition-id matcher tests have been removed. This matcher is
already removed in a newer GumTree version.

ibbem committed:
refactor: refactor the time dependent fields in `DiffNode`

This pattern should be more understandable than the previous array based
implementation. Furthermore, it is now much easier to add new time
dependent fields.

ibbem committed:
refactor: remove state cleanup in the diff parser

This is not necessary because the state is (and must be) only used for
parsing one variation diff. Furthermore, if this would be needed, it
would be buggy. In particular, this cleanup code is not executed in case
an exception is thrown.

ibbem committed:
refactor: use a switch for `GraphFormat`

This version will emit an error if `GraphFormat` is changed and doesn't
require an explicit exception for missing cases.

ibbem committed:
fix: use the correct child order in line graph exports

When importing line graphs, we consider the order of the edge lines as
the definite child ordering even if we export the child order using
`ChildOrderEdgeFormat`. However, the previous algorithm blindly exports
all `DiffType.NON` edges disregarding the fact the other edges might
need to be exported first. As the order of edge lines acts as a diff
(before edges act as deletions and after edges as insertions), we need
to compute a line diff of the child orders at both times to know which
edges can be exported as existing at both times (unchanged).

Note that some edges will be exported as inserted and deleted instead of
unmodified although both the child and the parent are the same because
line diffs cannot encode moves. Consumers of line graphs should not
depend on the fact which edges are exported as unchanged because the
diffing algorithm needs to apply heuristics to select these edges.

ibbem committed:
refactor: remove a duplicated `splitNode` implementation

ibbem committed:
refactor: simplify `GumTreeDiff.joinNode`

ibbem committed:
refactor: move `joinNode` into `DiffNode`

This makes it more discoverable and consistent with `DiffNode.split`.

ibbem committed:
test: activate incidentally solved won't fix tests

pmbittner commented on src/main/java/org/variantsync/diffdetective/variation/diff/construction/GumTreeDiff.java:
```diff
@@ -84,7 +92,18 @@ public static <B extends VariationNode<B, L>, L extends Label> DiffNode<L> diffU
         MappingStore matching = matcher.match(src, dst);
-        Assert.assertTrue(matching.has(src, dst));
+
+        // We assume that the root nodes are matched.
```
You don't just assume that, you enforce it here. Is it safe to do so?

pmbittner commented on src/main/java/org/variantsync/diffdetective/util/Assert.java:
```diff
@@ -86,7 +86,7 @@ public static void assertFalse(boolean condition, String errorMessage) {
 
     /** Throws {@link AssertionError} with {@code errorMessage} as error message. */
-    public static void fail(String errorMessage) {
+    public static <T> T fail(String errorMessage) {
```
absurd lol

pmbittner commented on src/main/java/org/variantsync/diffdetective/variation/diff/DiffNode.java:
```diff
@@ -51,32 +50,48 @@ public class DiffNode<L extends Label> implements HasNodeType {
-     * {@code getParent(time).getChildOrder(time).contains(this)}.
-     */
-    private DiffNode<L>[] parents = Cast.unchecked(Array.newInstance(DiffNode.class, 2));
+    private static class TimeDependent<L extends Label> {
```
This is non-trivial, so please add some documentation.

pmbittner replied:
From the name, I expected this to be a function `Time -> Value`. This function though is realized by having two fields (which is totally fine) in a DiffNode. For consistency with wording in our papers and my phd thesis, maybe `State` would be a better name? Or `ProjectedFields`? Or `AtTime`? I am unsure. I like the idea of the implementation but the naming is not trivial.

pmbittner commented on src/main/java/org/variantsync/diffdetective/variation/diff/DiffNode.java:
```diff
@@ -267,7 +279,7 @@ public void drop(Time time) {
      */
     public int indexOfChild(final DiffNode<L> child, Time time) {
-        return children[time.ordinal()].indexOf(child);
+        return timeDependent(time).children.indexOf(child);
```
Regarding the naming issue, one could ask here: What name would read well? How would be easiest discovered with IntelliSense?

`at(time).children....`?

`children.at(time)...` / `childrenAt(time)...`?

I like the second term but it breaks with the existing abstraction of `TimeDependent` or would require quite some additional boilderplate.

pmbittner commented on src/main/java/org/variantsync/diffdetective/variation/diff/DiffNode.java:
```diff
@@ -51,32 +50,48 @@ public class DiffNode<L extends Label> implements HasNodeType {
-     * {@code getParent(time).getChildOrder(time).contains(this)}.
-     */
-    private DiffNode<L>[] parents = Cast.unchecked(Array.newInstance(DiffNode.class, 2));
+    private static class TimeDependent<L extends Label> {
```
I appreciate this refactoring! Good idea. :)

ibbem committed:
docs: note that the diff parser does not require stacks

ibbem committed:
docs: add a disclaimer to `LineGraphImport`

ibbem committed:
test: create tests for `GumTreeDiff.improveMatching`

ibbem committed:
fix: respect trailing lines when matching variation trees

ibbem committed:
fix: let GumTreeDiff update the line numbers in the labels

ibbem committed:
fix: assign `after` line numbers in GumTreeDiff

ibbem committed:
fix: don't join syntactically different `DiffNode`s

ibbem commented:
In addition to addressing your feedback, I also discovered some bugs and added some tests.

There is one issue that I didn't address: The `GumTreeDiff` doesn't set the line numbers in the `DiffLinesLabel` correctly. However, although this is a bug in all current usages, the implementation is generic in the label and thus cannot generically merge/split labels. There are two possible fixes I can imagine:
- Add merge (and split) function arguments similarly as I did in `DiffNode.join` to fix this issue.
- Add merge (and split) functions to the `Label` interface.
What do you think?

pmbittner commented:
I like the idea of adding merge and split functions to the label interface. Would that break anything? Do both functions have meaning for our current implementations?

ibbem commented:
I added such functions in ec82fd49b24161935b2833f88bea9890a8d5438f. I'm not sure about their names though.

Should we write tests for the line numbers in the `DiffLinesLabel`? Currently, there are only tests for the line numbers in `DiffNode`.

pmbittner commented:
I clicked on the commit hash link in your comment and made some comments on that commit. Unfortunately, my comments do not appear here, so here are links to them:

- [x] https://github.com/VariantSync/DiffDetective/commit/ec82fd49b24161935b2833f88bea9890a8d5438f#r167331384
- [x] https://github.com/VariantSync/DiffDetective/commit/ec82fd49b24161935b2833f88bea9890a8d5438f#r167331046

pmbittner commented:
There seems to be a type error making CI fail. After fixing that, feel free to merge.
